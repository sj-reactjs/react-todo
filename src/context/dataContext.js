import React, { createContext, useState } from 'react'
import { defaultData } from './../mock/initial-data'

export const TodoDataContext = createContext()

function DataContextProvider(props) {
    const { children } = props;

    const [todoData, setTodoData] = useState(defaultData)
    const [formMode, setformMode] = useState('add')
    const [formData, setFormData] = useState({})

    function addTodo(todo) {
        setTodoData(currentTodo => {
            return [
                ...currentTodo,
                todo
            ]
        })
    }

    function removeTodo(id) {
        setTodoData(currentTodo => currentTodo.filter(todo => todo.id !== id))
    }

    function updateTodo(updatedTodo) {
        setTodoData((currentTodo) => {
            return currentTodo.map(todo => {
                if (todo.id === updatedTodo.id) {
                    return {
                        ...todo,
                        ...updatedTodo,
                    }
                }
                return todo
            })
        })
    }

    function setFormStage(stage = 'add', todoId) {
        switch (stage) {
            case 'edit':
                const itemToEdit = todoData.filter(({ id }) => `${todoId}` === `${id}`)
                if (itemToEdit.length) {
                    setformMode('edit')
                    setFormData(itemToEdit[0])
                }
                break;
            default:
                setformMode('add')
                setFormData({})
                break;
        }
    }

    return (
        <TodoDataContext.Provider value={{
            formMode,
            formData,
            todoData,
            add: addTodo,
            remove: removeTodo,
            update: updateTodo,
            setStage: setFormStage
        }}>
            {children}
        </TodoDataContext.Provider>
    )
}

export { DataContextProvider }
