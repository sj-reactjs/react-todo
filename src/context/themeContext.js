import React, { createContext, useState } from 'react'

const ThemeContext = createContext()

function ThemeContextProvider(props) {
    const { children } = props;
    const defaultTheme = 'dark'
    const [theme, setTheme] = useState(defaultTheme)

    function toggeleTheme() {
        setTheme(currentTheme => currentTheme === 'dark' ? 'light' : 'dark')
    }

    return (
        <ThemeContext.Provider value={{
            theme,
            changeTheme: toggeleTheme
        }}>
            <div className={`todo-app theme-${theme}`}>
                {children}
            </div>
        </ThemeContext.Provider>
    )
}

export { ThemeContextProvider, ThemeContext }
