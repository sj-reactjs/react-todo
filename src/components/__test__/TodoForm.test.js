import { render, screen } from '@testing-library/react'
import { TodoForm } from './../index'
import { ThemeContextProvider } from './../../context/themeContext'
import { DataContextProvider } from './../../context/dataContext'


describe(`Footer component`, () => {

    it(`render`, () => {
        const { asFragment, container } = render(
            <DataContextProvider>
                <ThemeContextProvider>
                    <TodoForm />
                </ThemeContextProvider>
            </DataContextProvider>
        )
        expect(asFragment()).toMatchSnapshot()

        const taskField = container.querySelector('input[id="task"]')
        if (taskField) {
            taskField.innerText = "new field"
        }
        const descriptionField = container.querySelector('input[id="description"]')
        descriptionField.innerText = "new desc"

        const selectField = container.querySelector('select[name="status"]')
        selectField.value = "in-progress"

        const submitButton = container.querySelector('button[type="submit"]')

        if (submitButton) {
            submitButton.click()
        }

        const cancelButton = container.querySelector('button[type="button"]')
        if (cancelButton) {
            cancelButton.click()
        }
    })
})