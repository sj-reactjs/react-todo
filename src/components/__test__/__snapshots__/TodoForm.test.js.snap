// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component render 1`] = `
<DocumentFragment>
  <div
    class="todo-app theme-dark"
  >
    <div>
      <form>
        <div>
          <div>
            <label
              for="task"
            >
              Task
            </label>
          </div>
          <input
            id="task"
            name="task"
            placeholder="Enter Task"
            type="text"
            value=""
          />
        </div>
        <div>
          <div>
            <label
              for="description"
            >
              Description
            </label>
          </div>
          <input
            id="description"
            name="description"
            placeholder="Description"
            value=""
          />
        </div>
        <div>
          <div>
            <label
              for="status"
            >
              Status
            </label>
          </div>
          <select
            id="status"
            name="status"
          >
            <option
              disabled=""
              value=""
            >
              Select Status
            </option>
            <option
              value="not-start"
            >
              Not Start
            </option>
            <option
              value="in-progress"
            >
              In Progress
            </option>
            <option
              value="complete"
            >
              Complete
            </option>
          </select>
        </div>
        <div>
          <button
            type="submit"
          >
            Submit
          </button>
          <button
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
